MTV-1897 Update performance recommendations for warm migration (#625)
kubev2v/forklift-documentation: documentation/modules/mtv-performance-recommendation.adoc

esxcli network ip interface tag add -t vSphereBackupNFC -i vmk2
----

[id="mtv-warm-migration-recommendations_{context}"]
== Warm migration recommendations

.Migrate up to 400 disks in parallel

Testing involved migrating 200 VMs, with 2 disks each using 8 ESX hosts. All VMs were distributed evenly over the ESX hosts, 25 VMs per ESX host. During the tests, the ESX hosts monitored the total connections of `DataTransfer`. The results show each ESX host had 50 open connections for disk transfer.

.High migration rate vs. total migration disks

Testing successfully performed parallel disk migrations with 200, 300, and 400 disks. There was a slight decrease in the pre-copy migration rate, approximately 25%, between the tests migrating 200 disks and those migrating 300 and 400 disks.

Therefore, we recommend performing batches of 300-400 parallel disk migrations, instead of 200 or fewer, only if a drop of 25% in precopy speed does not affect your cutover planning.

.Setting cutover time

> To reduce the overall time of warm migrations, we recommend setting the cutover right after the migration plan is started. This will cause the precopy phase of all VMs to run in parallel, subject to the setting of `MAX_VM_INFLIGHT`. This recommendation is valid no matter how many VMs are in the migration plan.

Reviewer: ```suggestion
To reduce the overall time of warm migrations, Red Hat recommends setting the cutover right after the migration plan is started. This will cause the precopy phase of all VMs to run in parallel, subject to the setting of `MAX_VM_INFLIGHT`. This recommendation is valid no matter how many VMs are in the migration plan.
```

.Increase precopy intervals between snapshots

> If you are creating many migration plans with a single VM and have enough time between the migration start and the cutover on the next day, we recommend increasing the value of the `controller_precopy_interval` parameter to between 120 and 240 minutes, inclusive. This will reduce the total number of snapshots and disk transfers per VM before the cutover.

Reviewer: ```suggestion
If you are creating many migration plans with a single VM and have enough time between the migration start and the cutover on the next day, Red Hat recommends increasing the value of the `controller_precopy_interval` parameter to between 120 and 240 minutes, inclusive. This will reduce the total number of snapshots and disk transfers per VM before the cutover.
```

* *Maximum cold migration transfer rate of the multiple VMs (single disk) from a single ESXi host: 294 MiB/s (concurrent migration of 30 VMs, 35/50 GiB used, from Single ESXi)*
* *Maximum cold migration transfer rate of the multiple VMs (single disk) from multiple ESXi hosts: 1173MB/s (concurrent migration of 80 VMs, 35/50 GiB used, from 8 ESXi servers, 10 VMs from each ESXi)*

[id="mtv-warm-migration-recommendations_{context}"]
== Warm migration recommendations

> The following are recommendations that are specific to warm migrations:

Reviewer 2: ```suggestion
The following recommendations are specific to warm migrations:
```

[id="mtv-warm-migration-recommendations-disks-in-parallel_{context}"]
=== Migrate up to 400 disks in parallel

> Testing involved migrating 200 VMs in parallel, with 2 disks each using 8 ESX hosts, for a total of 400 disks. No tests were run on migration plans migrating over 400 disks in parallel, so it is not recommended to migrate over this number of disks in parallel.

Reviewer 2: ```suggestion
Testing involved migrating 200 VMs in parallel, with 2 disks each using 8 ESXi hosts, for a total of 400 disks. No tests were run on migration plans migrating over 400 disks in parallel, so it is not recommended to migrate over this number of disks in parallel.
```

Be careful with your formatting of ESX and ESXi, be sure you are using the correct version

Author: I checked with David, who confirmed it should be "ESX".

Reviewer 3: I believe it should be ESXi. ESX is old discontinued technology.  We are working with ESXi check here https://storware.eu/blog/vmware-esx-vs-esxi-overview-of-key-differences/

[id="mtv-warm-migration-recommendations-effects-of-precopy-speed_{context}"]
=== Migrate up to 200 disks in parallel for the fastest rate

Testing was successfully performed on parallel disk migrations with 200, 300, and 400 disks. There was a decrease in the precopy migration rate, approximately 25%, between the tests migrating 200 disks and those migrating 300 and 400 disks.

> Therefore, it is recommended to perform parallel disk migrations in groups of 200 or fewer, instead of 300-400 disks, unless a drop of 25% in precopy speed does not affect your cutover planning.

Reviewer 2: ```suggestion
Therefore, it is recommended to perform parallel disk migrations in groups of 200 or fewer instead of 300 to 400 disks, unless a decline of 25% in precopy speed does not affect your cutover planning.
```

[id="mtv-warm-migration-recommendations-set-cutover-time_{context}"]
=== When possible, set cutover time to be immediately after a migration plan starts

> To reduce the overall time of warm migrations, it is recommended to set the cutover to occur immediately after the migration plan is started. This causes {project-short} to run only one precopy per VM. This recommendation is valid, no matter how many VMs are in the migration plan.

Reviewer 3: Could you Bold the "only one" ? 
I want it be clear - it is the main benifit

> To reduce the overall time of warm migrations, it is recommended to set the cutover to occur immediately after the migration plan is started. This will cause the precopy phase of all VMs to run in parallel, subject to the setting of `MAX_VM_INFLIGHT`. This recommendation is valid no matter how many VMs are in the migration plan.

Reviewer 2: ```suggestion
To reduce the overall time of warm migrations, it is recommended to set the cutover to occur immediately after the migration plan is started. This will cause the precopy phase of all VMs to run in parallel, subject to the setting of `MAX_VM_INFLIGHT`. This recommendation is valid, no matter how many VMs are in the migration plan.
```

[id="mtv-warm-migration-recommendations-increase-precopy-intevals_{context}"]
=== Increase precopy intervals between snapshots

If you are creating many migration plans with a single VM and have enough time between the migration start and the cutover, increase the value of the `controller_precopy_interval` parameter to between 120 and 240 minutes, inclusive. The longer setting will reduce the total number of snapshots and disk transfers per VM before the cutover.

[id="mtv-performance-max-values-warm_{context}"]
== Maximum values tested for warm migrations
* Maximum number of ESX hosts tested: 8
* Maximum number of worker nodes: 12
* Maximum number of VMs in a single migration plan: 200
* Maximum number of total parallel disk transfers: 400, with 200 VMs, 6 ESXs, and a transfer rate of 667 MB/s

> * Maximum number of number of disks on a single VM migrated: 3

Reviewer 2: ```suggestion
* Maximum number of disks on a single VM migrated: 3
```

A *number of numbers* LOL!!!

* *Maximum cold migration transfer rate of the multiple VMs (single disk) from a single ESXi host: 294 MiB/s (concurrent migration of 30 VMs, 35/50 GiB used, from Single ESXi)*
* *Maximum cold migration transfer rate of the multiple VMs (single disk) from multiple ESXi hosts: 1173MB/s (concurrent migration of 80 VMs, 35/50 GiB used, from 8 ESXi servers, 10 VMs from each ESXi)*

[id="mtv-warm-migration-recommendations_{context}"]
== Warm migration recommendations

The following are recommendations that are specific to warm migrations:

[id="mtv-warm-migration-recommendations-disks-in-parallel_{context}"]
=== Migrate up to 400 disks in parallel

> Testing involved migrating 200 VMs, with 2 disks each using 8 ESX hosts. All VMs were distributed evenly over the ESX hosts, 25 VMs per ESX host. During the tests, the ESX hosts monitored the total connections of `DataTransfer`. The results show each ESX host had 50 open connections for disk transfer.

Reviewer 3: I feel the wording here is less customer-friendly. Currently, it sounds like a test report. Please consider rewording.something like " Testing involved migrating 200 VMs, with 2 disks each using 8 ESX hosts. All VMs were distributed evenly over the ESX hosts, 25 VMs per ESX host - 400 disks TotallAnd say that we cant recommend going higher than what we tested... we don't know what will be the impact.
 Remove this - to technical: "During the tests, the ESX hosts monitored the total connections of DataTransfer. The results show each ESX host had 50 open connections for disk transfer."  @duduvaa  what do you think

[id="mtv-warm-migration-recommendations-effects-of-precopy-speed_{context}"]
=== Precopy speed drops for parallel disk migrations of over 200 disks

Testing was successfully performed on parallel disk migrations with 200, 300, and 400 disks. There was a slight decrease in the precopy migration rate, approximately 25%, between the tests migrating 200 disks and those migrating 300 and 400 disks.

Therefore, it is recommended to perform parallel disk migrations in groups of 300-400 disks, instead of 200 or fewer, only if a drop of 25% in precopy speed does not affect your cutover planning.

[id="mtv-warm-migration-recommendations-set-cutover-time_{context}"]

> === Set cutover time to be immediately after a migration plan starts

Reviewer 3: Change to "When Possible Set the  cutover time...
As not always cut over can be set like this and customers may have struct maintenance windows

[id="mtv-warm-migration-recommendations-increase-precopy-intevals_{context}"]
=== Increase precopy intervals between snapshots

If you are creating many migration plans with a single VM and have enough time between the migration start and the cutover, increase the value of the `controller_precopy_interval` parameter to between 120 and 240 minutes, inclusive. The longer setting will reduce the total number of snapshots and disk transfers per VM before the cutover.

[id="mtv-performance-max-values-warm_{context}"]
== Maximum values tested for warm migrations
* Maximum number of ESX hosts tested: 8
* Maximum number of worker nodes: 12
* Maximum number of VMs in a single migration plan: 200
* Maximum number of total parallel disk transfers: 400, with 200 VMs, 6 ESXs, and a transfer rate of 667 MB/s
* Maximum number of number of disks on a single VM migrated: 3
* Maximum number of parallel disk transfers per ESX host: 68

> * Maximum transfer rate observed of a single disk with no concurrent migrations: 76.5 MB/s

Reviewer 3: Can you bold this line and the 2 after it Same as we did for cold?

* Maximum transfer rate observed of multiple disks from a single ESX host: 253 MB/s (concurrent migration of 10 VMs, 1 disk each, 35/50 GiB used per disk)
* Total transfer rate observed of multiple disks (210) from 8 ESX hosts: 802 MB/s (concurrent migration of 70 VMs, 3 disks each, 35/50 GiB used per disk)

> For additional details on performance, see xref:mtv-performance-addendum_mtv[{project-short} performance addendum].

Reviewer 3: Delete this line, In the addendum there is only cold migration results

Author: @dgur Do you want me to delete it or should I return it to the end of "Maximum values tested for cold migrations" ?

* *Maximum cold migration transfer rate of the multiple VMs (single disk) from a single ESXi host: 294 MiB/s (concurrent migration of 30 VMs, 35/50 GiB used, from Single ESXi)*
* *Maximum cold migration transfer rate of the multiple VMs (single disk) from multiple ESXi hosts: 1173MB/s (concurrent migration of 80 VMs, 35/50 GiB used, from 8 ESXi servers, 10 VMs from each ESXi)*

[id="mtv-performance-max-values-warm_{context}"]

> == Maximum values tested for warm migrations

Reviewer: I see that the parameters are complicated. Is a table possible in this scenario? The values will be clearly visualized if they are put into a table. Leaving it to you to check for feasibility.

Author: @shdeshpa07 Thanks, but I'm going to keep the same format as the preceding section.
